### DISCUSSION TITLE
How to substract a Ball.stl from a Cube.stl and save it as stl again?

### DISCUSSION THREAD

**Discussion by Boehm92** [jimy-byerley/pymadcad#22] [Q&A] [answered]
Hi, 
I'm new here. I found this repo and would love to use it in my Ph.D. research. I would like to simply subtract two stl objects(e.g. cube, ball) from each other, visualize it and save it again as stl file.

I found an example code in the issue section and tried it with the supplied files. If I delete the mc.segmentation code it works for me, else I get ('NoneType' object has no attribute 'points').

I would be thankful for a simple example.

Kind Regards 
Stefan

**Comment by jimy-byerley** [comment, marked as answer]
Welcome there @Boehm92 !
it you only need to reexport to `.stl` then you don't need the segmentation step, the segmentation only groups triangles by guessing soft and sharp regions in the mesh, it is helpful to have a nice visualization or to perform engineering operations by designating groups. but it you only need to reexport then it's pointless since the STL format doesn't cary groups but only triangles.
Something as simple as this should work:

```python
import madcad as mc

a = mc.read('file1.stl')
b = mc.read('file2.stl')
a.mergeclose()   # don't forget to merge close points from STL files, as they provide unconnected triangles, we have to connect it again before boolean operations
b.mergeclose()
result = mc.difference(a, b)

mc.show([result])
mc.write('result.stl')
```

What is the example you are refering to which is not working properly ? so I can take a look

**Reply by Boehm92** [reply]
Thanks a lot, it works! This brings me a big step further.

The example I used before was your code of the issue "Boolean operations failing in case of overlapping faces #8". I had some problems with my virtual environment, your example from the issue wasn't the problem.

**Reply by jimy-byerley** [reply]
ah ^^ I'm glad to hear this since it won't be a bug I would have to fix !
